### PR TITLE
914 [Feature] - "Add remaining CRUD endpoints to SST"

### LINKED ISSUE #914: [FEATURE]: Add the Rest of the CRUD Endpoints to SST
### Feature Name

Add the rest of the CRUD API endpoints to SST

### Your Name

Daniel Wu

### Description

Serverless Stack Toolkit (SST) handles development and deployment of serverless stacks, simplifying the development experience for lambdas. More info here [https://www.youtube.com/watch?v=JY_d0vf-rfw](https://www.youtube.com/watch?v=JY_d0vf-rfw) and here [https://docs.sst.dev/](https://docs.sst.dev/).

Endpoint routes to add:
- POST /trainspace (creates a trainspace and returns the id)
- GET /trainspace/{id} (gets trainspace information given the id)
- GET /trainspace (lists all trainspaces of a user)
- DELETE /trainspace/{id} (deletes a trainspace given its id)
- DELETE /dataset/user/{type}/{filename} (deletes a user uploaded dataset)

All code will be in the `serverless` directory.

### PR DESCRIPTION
**Pull request by alantao912**
# Added remaining trainspace CRUD endpoints to SST

**What user problem are we solving?**
Trainspace creation, deletion, management, and deletion of user uploaded datset.

**What solution does this PR provide?**
Allowing users to create new, retrieve specific, retrieve all, delete specific trainspace, and delete specific user uploaded dataset using AWS Lambda.

**Testing Methodology**
Tested API endpoints on locally with postman using a developer account. 

Created new trainspace, retrieved it, created several new trainspaces, retrieved all, deleted trainspace, attempted (and successfully failed) to retrieve previously deleted trainspace, and deleted user uploaded data successfully.

**Any other considerations**
- More thorough testing may be needed for deleting datasets for S3 buckets, haven't checked false negative case. No exceptions were thrown however.
- May need to improve exception messages, and exception detection.

### CODE REVIEW

File: serverless/packages/functions/src/datasets/user/delete_userdata.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import { S3Client, ListObjectsV2Command } from "@aws-sdk/client-s3";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+    return {
+        statusCode: 404,
+        body: JSON.stringify({ message: "Not Found" }),
```
Comment (karkir0003) [inline serverless/packages/functions/src/datasets/user/delete_userdata.ts]: I'm confused here. Why do we always toss a 404?

**Comment by alantao912** [comment]
Sept 23rd. Latest

File: serverless/packages/functions/src/trainspace/get_all_trainspace.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            KeyConditionExpression: "uid = :uid",
+            ExpressionAttributeValues: {
+                ":uid" : 
+                { 
+                    S: user_id
+                }
+            }
+        });
+
+        const results = await docClient.send(getCommand);
```
Comment (karkir0003) [inline serverless/packages/functions/src/trainspace/get_all_trainspace.ts]: From my understanding Dynamo DB will paginate responses when a Get/query operation is called. We might have to use LastEvaluatedKey to get all entries/trainspaces corresponding to a given user id. I think we should explore how to use LastEvaluatedKey in order to ensure the get all trainspace endpoint works correctly

https://w11i.me/paging-dynamodb-queries-with-typescript-generators

https://stackoverflow.com/questions/44589967/how-to-fetch-scan-all-items-from-aws-dynamodb-using-node-js

@alantao912

**Review by karkir0003 — COMMENTED** [review]
overall looks good to me. left a comment

File: serverless/packages/functions/src/trainspace/get_all_trainspace.ts (hunk 2 above)
Comment (alantao912) [reply serverless/packages/functions/src/trainspace/get_all_trainspace.ts]: I will look into this. Thanks for bringing it to my attention.

**Review by karkir0003 — COMMENTED** [review]
also, can you attach a screen recording to show that the api endpoint works and data is updated on trainspace dynamodb correctly?

**Comment by alantao912** [comment]
> also, can you attach a screen recording to show that the api endpoint works and data is updated on trainspace dynamodb correctly?

https://youtu.be/aMiK-I8cpqs

**Review by karkir0003 — COMMENTED** [review]
Looks good to me. Thanks for the awesome work on this. @dwu359 will review your pr and leave his comments.

**Comment by karkir0003** [comment]
@alantao912 if you've addressed all the PR comments and tested your changes using the method outlined in your video, let me know and I can mark your PR as approved to unblock

**Review by dwu359 — CHANGES_REQUESTED** [review]
There were a few typing issues when I reviewed your PR, make sure to add the correct packages to the package.json via yarn or the dlp-cli, and make sure to add the @types/.. equivalents if that exists. 

Also make sure that your endpoints check if the parameters needed do exist and are formatted correctly in the request. For example, the create trainspace endpoint should check if there is a name, data_source, dataset_data, parameters_data, review_data, etc. It should also check if the data_source param is IMAGE, TABULAR, etc and the data values have the correct format (can check the required format in frontend/src/features/Train/... and look for types files). For POST endpoints, the params should be part of the body. For GET endpoints the params should be query params.

It's a good start though, make sure to ask questions especially about the above if you are confused about anything.

**Comment by karkir0003** [comment]
@alantao912 there are some merge conflicts to be resolved?

File: serverless/packages/functions/src/trainspace/create_trainspace.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+import TrainspaceData from './trainspace-data';
+import { v4 as uuidv4 } from 'uuid';
+import { DynamoDBClient, PutItemCommand } from '@aws-sdk/client-dynamodb'; //@aws-sdk/client-dynamodb
+import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
+import { DATA_SOURCE_ARR } from '../../../../../frontend/src/features/Train/constants/trainConstants';
```
Comment (karkir0003) [inline serverless/packages/functions/src/trainspace/create_trainspace.ts]: Are you sure that this array can only come from this directory? @farisdurrani can you take a look?

Comment (farisdurrani) [reply serverless/packages/functions/src/trainspace/create_trainspace.ts]: Ooffff, this is bad. You shouldn't be importing stuff from the /frontend or /backend directories. We need to consider them separate microservices because they'll be Dockerized separately. I'd recommend making a copy of the data source array in /trainspace somewhere

Comment (karkir0003) [reply serverless/packages/functions/src/trainspace/create_trainspace.ts]: @alantao912

Comment (alantao912) [reply serverless/packages/functions/src/trainspace/create_trainspace.ts]: Okay, sure.

**Comment by karkir0003** [comment]
@alantao912 can you fix the merge conflicts and also test the endpoints?

**Comment by alantao912** [comment]
> @alantao912 can you fix the merge conflicts and also test the endpoints?

No merge conflicts. Fast-forwarded and pushed.

**Comment by karkir0003** [comment]
> > @alantao912 can you fix the merge conflicts and also test the endpoints?
> 
> No merge conflicts. Fast-forwarded and pushed.

did you test the endpoints @alantao912

File: serverless/packages/functions/src/trainspace/create_trainspace.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        case "CLASSICAL_ML":
+
+            return null;
+        case "OBJECT_DETECTION":
+
+            return null;
+        default:
+            return null;
+    }
+}
```
Comment (karkir0003) [inline serverless/packages/functions/src/trainspace/create_trainspace.ts]: potential code smell. Anyway we can improve this code @alantao912 @farisdurrani @dwu359

File: serverless/stacks/AppStack.ts [hunk 5]
```diff
@@ -55,6 +55,32 @@ export function AppStack({ stack }: StackContext) {
         "packages/functions/src/datasets/user/all_of_type.handler",
       "GET /datasets/user/{type}/{filename}/columns":
         "packages/functions/src/datasets/user/columns.handler",
+      "DELETE /dataset/user/{type}/{filename}" :
+        "packages/functions/src/datasets/user/delete_userdata.handler",
+      "POST /trainspace": {
+        function: {
+          handler: "packages/functions/src/trainspace/create_trainspace.handler",
+          permissions: ["dynamodb"]
```
Comment (karkir0003) [inline serverless/stacks/AppStack.ts]: what exact permissions does SST get from passing in "dynamodb"? Is principle of least privilege enforced here? @alantao912

Comment (alantao912) [reply serverless/stacks/AppStack.ts]: I just looked into this, there are more granular permissions such as dynamodb:PutItem, dynamodb:GetItem, etc. I will implement this.

File: serverless/packages/functions/src/trainspace/create_trainspace.ts (hunk 4 above)
Comment (alantao912) [reply serverless/packages/functions/src/trainspace/create_trainspace.ts]: Should I have used polymorphism here instead ? I thought about it but I don't think that would get around having to check the string

Comment (karkir0003) [reply serverless/packages/functions/src/trainspace/create_trainspace.ts]: lets keep this for now but might wanna refactor later

**Comment by alantao912** [comment]
@karkir0003 Hey, narrowed permissions on the SST endpoints and tested with the same method as before. Should be good to go.

File: serverless/packages/functions/src/trainspace/create_trainspace.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+import TrainspaceData from './trainspace-data';
+import { v4 as uuidv4 } from 'uuid';
+import { DynamoDBClient, PutItemCommand } from '@aws-sdk/client-dynamodb'; //@aws-sdk/client-dynamodb
```
Comment (karkir0003) [inline serverless/packages/functions/src/trainspace/create_trainspace.ts]: some commented out code here. please clean up @alantao912

File: serverless/stacks/AppStack.ts [hunk 7]
```diff
@@ -11,6 +11,7 @@ export function AppStack({ stack }: StackContext) {
       ),
     },
   });
+  /*
```
Comment (karkir0003) [inline serverless/stacks/AppStack.ts]: why is this comment here? @alantao912

**Review by karkir0003 — CHANGES_REQUESTED** [review]
minor comments/nits

File: serverless/stacks/AppStack.ts (hunk 7 above)
Comment (alantao912) [reply serverless/stacks/AppStack.ts]: I had to remove some code that was causing issues.

Comment (karkir0003) [reply serverless/stacks/AppStack.ts]: got it! lets fix this to have idempotency as discussed in dm

**Review by karkir0003 — APPROVED** [review]
good work on this @alantao912 . approved!

please create a github issue for the idempotency thing we discussed earlier today

**Review by dwu359 — COMMENTED** [review]
I think it would be better if each trainspace type was handled by it's own crud api endpoints rather than having a conglomerate trainspace crud endpoints. Our current way of approaching this seems to be leading to the massive switch statement, giant functions, etc, and from looking at the current implementation, it seems like each trainspace type seems to have fundamentally different needs when it comes to implementing the CRUD endpoints for them. I think a CRUD api endpoint set each for tabular, image, classical, etc would be better. @karkir0003 what do you think

File: serverless/packages/functions/src/trainspace/trainspace-data.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        this.name = name;
+        this.parameters_data = parameters_data;
+        this.review_data = review_data;
+    }
+
+    getDateString() : string {
+        return Date.now().toString();
+    }
+
+    convertToDynamoItemInput(tableName: string) : PutItemCommandInput {
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/trainspace-data.ts]: I don't think a class is necessary for this, since the convertToDynamoItemInput can be generalized so that it applies to any dynamo item. What's left of the class would then just be the data, which would defeat the purpose of having the class in the first place

File: serverless/packages/functions/src/datasets/user/delete_userdata.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
+import { S3Client, DeleteObjectCommand } from "@aws-sdk/client-s3";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
```
Comment (dwu359) [inline serverless/packages/functions/src/datasets/user/delete_userdata.ts]: Is there endpoints for create and get user data also?

**Comment by karkir0003** [comment]
> I think it would be better if each trainspace type was handled by it's own crud api endpoints rather than having a conglomerate trainspace crud endpoints. Our current way of approaching this seems to be leading to the massive switch statement, giant functions, etc, and from looking at the current implementation, it seems like each trainspace type seems to have fundamentally different needs when it comes to implementing the CRUD endpoints for them. I think a CRUD api endpoint set each for tabular, image, classical, etc would be better. @karkir0003 what do you think

fair point @dwu359 but what would the diff between these crud endpoints look like?

**Comment by dwu359** [comment]
> > I think it would be better if each trainspace type was handled by it's own crud api endpoints rather than having a conglomerate trainspace crud endpoints. Our current way of approaching this seems to be leading to the massive switch statement, giant functions, etc, and from looking at the current implementation, it seems like each trainspace type seems to have fundamentally different needs when it comes to implementing the CRUD endpoints for them. I think a CRUD api endpoint set each for tabular, image, classical, etc would be better. @karkir0003 what do you think
> 
> fair point @dwu359 but what would the diff between these crud endpoints look like?

A set of CRUD endpoints for tabular, a set of CRUD endpoints for image, a set of CRUD endpoints for classical ml. Basically just a separation of concerns

**Comment by karkir0003** [comment]
like treat trainspace types differently? dont we already segregate tabular train from say image train

**Comment by karkir0003** [comment]
@alantao912 thoughts?

**Comment by karkir0003** [comment]
what makes img trainspace different from say a tabular trainspace?

**Comment by dwu359** [comment]
Yes. Although all trainspaces are still stored in the same trainspace table. We would still need a get all trainspaces endpoint for the dashboard and a delete trainspace by id for deleting from the dashboard. So I guess it's just the create endpoints that need to get separated then

**Comment by dwu359** [comment]
> what makes img trainspace different from say a tabular trainspace?

Mainly different json structures, which may require different ways of handling such data

**Comment by alantao912** [comment]
> @alantao912 thoughts?

I don't think there would need to be a _complete set_ of new CRUD endpoints for each type of trainspace. There only need to be different handling for creating new trainspaces. Deletion, reading, should not vary between data sources.

File: serverless/packages/functions/src/datasets/user/delete_userdata.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
+import { S3Client, DeleteObjectCommand } from "@aws-sdk/client-s3";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+    if (
+        event &&
+        event.pathParameters &&
```
Comment (karkir0003) [inline serverless/packages/functions/src/datasets/user/delete_userdata.ts]: why is this file called delete_userdata.ts. I'm confused on what the code here has to do with user data @alantao912

**Review by karkir0003 — COMMENTED** [review]
added another comment

File: serverless/packages/functions/src/trainspace/delete_trainspace.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+
+import { DynamoDBClient, DeleteItemCommand } from '@aws-sdk/client-dynamodb';
+import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/delete_trainspace.ts]: Good find, didn't know about this handy wrapper library before

File: serverless/packages/functions/src/trainspace/delete_trainspace.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+
+import { DynamoDBClient, DeleteItemCommand } from '@aws-sdk/client-dynamodb';
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/delete_trainspace.ts]: You need to import the commands from `lib-dynamodb` instead of `client-dynamodb`, since you are using the document client instead of the dynamo client. This seems to be causing the typescript errors for other lambdas also

File: serverless/packages/functions/src/trainspace/delete_trainspace.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        
+        const client = new DynamoDBClient({});
+        const documentClient = DynamoDBDocumentClient.from(client);
+        
+        const command = new DeleteItemCommand({
+            TableName : "trainspace",
+            Key :
+            {
+                "trainspace_id" :
+                {
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/delete_trainspace.ts]: document client should automatically marshall input according to the docs: [https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-lib-dynamodb/](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-lib-dynamodb/)

File: serverless/stacks/AppStack.ts [hunk 14]
```diff
@@ -69,9 +82,6 @@ export function AppStack({ stack }: StackContext) {
       api.getFunction("GET /datasets/user/{type}")?.functionName ?? "",
     GetUserDatasetColumnsFunctionName:
       api.getFunction("GET /datasets/user/{type}/{filename}/columns")
-        ?.functionName ?? "",
-    TrainspaceRunTableName: {
-      value: trainspaceRunTable.tableName
-    }
+        ?.functionName ?? ""
```
Comment (dwu359) [inline serverless/stacks/AppStack.ts]: Include all endpoint function names as output here

File: serverless/packages/functions/src/dbutils/put_trainspace.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
+import { TrainStatus } from '../trainspace/constants';
+import { PutItemCommandInput } from "@aws-sdk/client-dynamodb";
+
+export function create_trainspace(trainspace_id: string, uid: string, data_source: string, dataset_data: object, name: string, parameters_data: object, review_data: string) : PutItemCommandInput | null
```
Comment (dwu359) [inline serverless/packages/functions/src/dbutils/put_trainspace.ts]: Hopefully not needed because of document client library

File: serverless/packages/functions/src/datasets/user/delete_user_uploaded_data.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
+import { APIGatewayProxyHandlerV2 } from "aws-lambda";
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
+import { S3Client, DeleteObjectCommand } from "@aws-sdk/client-s3";
+import parseJwt from "@dlp-sst-app/core/parseJwt";
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
```
Comment (dwu359) [inline serverless/packages/functions/src/datasets/user/delete_user_uploaded_data.ts]: rename files so that they are less verbose (e.g. `delete_user_uploaded_data` can just be `delete`)

Comment (dwu359) [reply serverless/packages/functions/src/datasets/user/delete_user_uploaded_data.ts]: actually delete_url would probably be a better name for this

**Review by dwu359 — CHANGES_REQUESTED** [review]
Looks better

File: serverless/packages/functions/src/datasets/user/delete_user_uploaded_data.ts (hunk 16 above)
Comment (karkir0003) [reply serverless/packages/functions/src/datasets/user/delete_user_uploaded_data.ts]: agreed

File: serverless/packages/functions/src/trainspace/get_all_trainspace.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+        do {
+            const getCommand: QueryCommand = new QueryCommand({
+                TableName: "trainspace",
+                IndexName: "uid",
+                KeyConditionExpression: "uid = :uid",
+                ExpressionAttributeValues: {
+                    ":uid" : 
+                    { 
+                        S: user_id
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/get_all_trainspace.ts:24]: Use lib-dynamodb's QueryCommand (not client-dynamodb) from DynamoDBClient for here also if possible

File: serverless/packages/functions/src/trainspace/create_image_trainspace.ts [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
+import { create_trainspace } from "../dbutils/put_trainspace";
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+    if (event) {
+        const uid: string = parseJwt(event.headers.authorization ?? "")["user_id"];
+        const eventBody = JSON.parse(event.body? event.body : "");
+
+        const trainspaceId = uuidv4();
+        const putCommandInput = create_trainspace(trainspaceId, uid, "IMAGE", eventBody['dataset_data'], eventBody['name'], eventBody['parameters_data'], eventBody['review_data']['notification_email']);
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/create_image_trainspace.ts]: creating the trainspace object shouldn't be generalized to a function, since the attributes for each trainspace can vary between trainspace types (image, tabular, etc)

File: serverless/packages/functions/src/trainspace/create_tabular_trainspace.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
+import { create_trainspace } from '../dbutils/put_trainspace';
+
+export const handler: APIGatewayProxyHandlerV2 = async (event) => {
+    if (event) {
+        const uid: string = parseJwt(event.headers.authorization ?? "")["user_id"];
+        const eventBody = JSON.parse(event.body? event.body : "");
+
+        const trainspaceId = uuidv4();
+        const putCommandInput = create_trainspace(trainspaceId, uid, "TABULAR", eventBody['default'], eventBody['name'],
```
Comment (dwu359) [inline serverless/packages/functions/src/trainspace/create_tabular_trainspace.ts]: same as above

File: serverless/packages/functions/src/dbutils/put_trainspace.ts [hunk 20]
```diff
@@ -0,0 +1,24 @@
+import { TrainStatus } from '../trainspace/constants';
+import { PutCommandInput } from "@aws-sdk/lib-dynamodb";
+
+export function create_trainspace(trainspaceId: string, user_id: string, dataSource: string, dataset_data: object, trainspaceName: string, parameters_data: object, reviewData: string) : PutCommandInput | null
+{
+    let output: PutCommandInput =
```
Comment (dwu359) [inline serverless/packages/functions/src/dbutils/put_trainspace.ts]: In the future, we plan to add other trainspace types such as audio, cnns, neural net, etc that may make this function to cumbersome to generalize

**Review by dwu359 — CHANGES_REQUESTED** [review]
A few minor changes before I do a full review. Also rename these files to be consistent with the rest of the sst code:
- create_image_trainspace -> create_image
- create_tabular_trainspace -> create_tabular
- delete_trainspace -> delete
- get_all_trainspace -> get_all

File: serverless/packages/functions/src/trainspace/create_image_trainspace.ts (hunk 18 above)
Comment (karkir0003) [reply serverless/packages/functions/src/trainspace/create_image_trainspace.ts:14]: whats stopping us from having separate trainspace creation functions? that way each trainspace we create has a predictable structure with some differences? rather than hard coding all props of trainspace down at sst? @danielwu

**Review by farisdurrani — APPROVED** [review]

**Comment by farisdurrani** [comment]
You're quite behind the nextjs branch. Just fix the conflicts and you should be good to go 👍 @alantao912

**Review by dwu359 — APPROVED** [review]
